Add way to color room characters

Delwing opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions

Add way to change color of room symbol through map context menu and Lua API
![image](https://user-images.githubusercontent.com/3740628/103467976-d4571300-4d54-11eb-8a83-130a153bbe10.png)
Map symbol picking is no longer modal as well.

#### Motivation for adding to Mudlet

More customization for maps of course!

#### Other info (issues closed, discussion etc)
closes #3988

Commit:
add way to color room characters

Commit:
accept new translations

Commit:
fix qmake build

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -11153,6 +11153,107 @@ int TLuaInterpreter::getRoomChar(lua_State* L)
+    }
+    int r = lua_tonumber(L, 2);
+    if (r < 0 || r > 255) {
+        lua_pushfstring(L, "setRoomCharColor: bad argument #2 type (red component value %d out of range (0 to 255)", r);
```
These are a nil + errormessage type case (value out of range rather than being the wrong type). It shouldn't throw a `lua_error(...)` IMO.

Delwing replied:
This is equally invalid as putting string actually.

vadi2 replied:
I agree, it's an error to misuse the API. There's no case where this would ever be valid, so it can't be nil+error

SlySven commented on src/TMap.cpp:
```diff
@@ -1224,6 +1224,12 @@ bool TMap::serialize(QDataStream& ofs, int saveVersion)
             }
             ofs << oldCharacterCode;
         }
+        if(mSaveVersion >= 21) {
```
Like wot CodeFactor sez...

Delwing replied:
Nutin.. :) Corrected.

SlySven commented on src/TMap.cpp:
```diff
@@ -2628,6 +2634,102 @@ void TMap::update()
+        }
+    }
+    switch (env) {
+    case 1:
```
If it were me I'd collapse these all down to single line `case`s - with the elements all padded out to 4-space columns (like the tab key will do in QT Creator if it is set to use only 4 spaces for indentation) like this:
```cpp
        switch (env) {
        case 1:     color = mpHost->mRed_2;             break;
        case 2:     color = mpHost->mGreen_2;           break;
        case 3:     color = mpHost->mYellow_2;          break;
        case 4:     color = mpHost->mBlue_2;            break;
        case 5:     color = mpHost->mMagenta_2;         break;
        case 6:     color = mpHost->mCyan_2;            break;
        case 7:     color = mpHost->mWhite_2;           break;
        case 8:     color = mpHost->mBlack_2;           break;
        case 9:     color = mpHost->mLightRed_2;        break;
        case 10:    color = mpHost->mLightGreen_2;      break;
        case 11:    color = mpHost->mLightYellow_2;     break;
        case 12:    color = mpHost->mLightBlue_2;       break;
        case 13:    color = mpHost->mLightMagenta_2;    break;
        case 14:    color = mpHost->mLightCyan_2;       break;
        case 15:    color = mpHost->mLightWhite_2;      break;
        case 16:    color = mpHost->mLightBlack_2;      break;
        default: //user defined room color
            if (!customEnvColors.contains(env)) {
                if (16 < env && env < 232) {
                    quint8 base = env - 16;
                    quint8 r = base / 36;
                    quint8 g = (base - (r * 36)) / 6;
                    quint8 b = (base - (r * 36)) - (g * 6);
    
                    r = r * 51;
                    g = g * 51;
                    b = b * 51;
                    color = QColor(r, g, b, 255);
                } else if (231 < env && env < 256) {
                    quint8 k = ((env - 232) * 10) + 8;
                    color = QColor(k, k, k, 255);
                }
                break;
            }
            color = customEnvColors[env];
        }
```

Delwing replied:
Changed.

Commit:
addSymbolToPixmapCache improve

SlySven commented on src/dlgRoomSymbol.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+    if (mpSymbols.size() <= 1) {
+        return lineEdit_roomSymbol->text();
+    } else {
```
:information_source: There is no need for the `else { ... }` as the `return` on the line before will mean that the formerly `else` code can be unindented one level - this is the coding anti-pattern that several *linters* call `else-after-return`... https://clang.llvm.org/extra/clang-tidy/checks/readability-else-after-return.html

Delwing replied:
This one slipped. Corrected.

SlySven commented on src/ui/room_symbol.ui:
```diff
@@ -0,0 +1,210 @@
+  <property name="windowTitle">
+   <string>Room symbol</string>
+  </property>
+  <layout class="QGridLayout" name="gridLayout">
```
I think you have some redundancy here - this `QGridLayout` only has one member..

Delwing replied:
Leftover after playing around with layout. Removed.

SlySven reviewed (commented):
I will need to build this before I can give a more fuller review...

Commit:
cr corrections

vadi2 committed:
Modernize design

vadi2 commented:
I added https://github.com/Delwing/Mudlet/pull/6 which improves on the design of the dialog:

![image](https://user-images.githubusercontent.com/110988/103474693-b914f380-4da6-11eb-847f-c0bd4a73f285.png)
![image](https://user-images.githubusercontent.com/110988/103474697-bb774d80-4da6-11eb-87aa-499180ddaa6d.png)

vadi2 commented on src/ui/room_symbol.ui:
```diff
@@ -0,0 +1,203 @@
+            <item>
+             <widget class="QLineEdit" name="lineEdit_roomSymbol">
+              <property name="inputMethodHints">
+               <set>Qt::ImhUrlCharactersOnly</set>
```
What's the reason for enabling this?

Delwing replied:
Some miscclick I guess.

vadi2 replied:
There is a little arrow button you can use to unset a property, instead of having it set to none

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3647,144 +3551,35 @@ void T2DMap::slot_setSymbol()
-                }
-            }
-        }
+        QString inputSymbol = QString();
```
```suggestion
        QString inputSymbol;
```

Delwing replied:
Actually remove whole line. Not needed.

vadi2 commented on src/TRoom.h:
```diff
@@ -146,6 +146,7 @@ class TRoom
     qreal max_x;
     qreal max_y;
     QString mSymbol;
+    QColor mSymbolColor = nullptr;
```
Just `QColor mSymbolColor;` is enough

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3647,144 +3551,35 @@ void T2DMap::slot_setSymbol()
+        auto symbolDialog = new dlgRoomSymbol(mpHost, this);
+        symbolDialog->init(usedSymbols, roomPtrsSet);
+        symbolDialog->show();
+        symbolDialog->raise();
```
You also want to set the `Qt::WA_DeleteOnClose` property so it gets deleted when closed to free the memory.

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3647,144 +3551,35 @@ void T2DMap::slot_setSymbol()
-            }
-        }
+        QString inputSymbol = QString();
+        auto symbolDialog = new dlgRoomSymbol(mpHost, this);
```
The new design allows spawning multiple dialogs at once, even on the same room, maybe not the best idea?

Delwing replied:
I can't figure out how custom lines dialog is done. It blocks map interactions, but does not block main window... This could be applied.
In theory it has `exec` call which make this window modal... but when I do set modality on symbol dialog it blocks main window as well.

vadi2 replied:
Remember the fix for multiple About buttons - same here, store a pointer and if the dialog is open, don't show another one.

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3647,144 +3551,35 @@ void T2DMap::slot_setSymbol()
-                itRoomPtr.next()->mSymbol = newSymbol;
-            }
+    } else {
+        // 8.0 is the maximum supported by all the Qt versions (>= 5.7.0) we
```
@SlySven this is now outdated as we use Qt 5.11 at minimum, does it need an update?

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -11153,6 +11153,107 @@ int TLuaInterpreter::getRoomChar(lua_State* L)
+    Host& host = getHostFromLua(L);
+    TRoom* pR = host.mpMap->mpRoomDB->getRoom(id);
+    if (!pR) {
+        lua_pushfstring(L, "setRoomCharColor: room with id %d does not exist", id);
```
This should not be an error but nil+msg instead, it should not break scripts execution to have gotten the room ID wrong

vadi2 commented on src/dlgRoomSymbol.cpp:
```diff
@@ -0,0 +1,219 @@
+#include <QPainter>
+#include "post_guard.h"
+
+dlgRoomSymbol::dlgRoomSymbol(Host* pHost, QWidget* pF) : QDialog(pF), mpHost(pHost)
```
Sidenote, we ought to give `pF` a much better name everywhere in the project.

Kebap replied:
Put a new issue for future reference, so it does not get buried with this PR merged: #6375

Delwing committed:
Merge pull request #6 from vadi2/room-symbol-color-selection-improve-layout

Improve room symbol dialog design

vadi2 commented on src/dlgRoomSymbol.cpp:
```diff
@@ -0,0 +1,219 @@
+    label_preview->setFont(font);
+}
+
+void dlgRoomSymbol::init(QHash<QString, unsigned int>& pSymbols, QSet<TRoom*>& pRooms)
```
Sidenote, I know this came from existing code, but `unsigned` here is an anti-pattern according to modern C++ guidelines: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-nonnegative

We should avoid it in the future and remove its use here. The more you know!

See also `Expects`: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i6-prefer-expects-for-expressing-preconditions

Kebap replied:
Put a new issue for future reference, so it does not get buried with this PR merged: #6376

vadi2 commented on src/dlgRoomSymbol.cpp:
```diff
@@ -0,0 +1,219 @@
+
+void dlgRoomSymbol::initInstructionLabel()
+{
+    if (mpSymbols.size() == 0) {
```
```suggestion
    if (mpSymbols.empty()) {
```

vadi2 commented on src/dlgRoomSymbol.cpp:
```diff
@@ -0,0 +1,219 @@
+{
+    auto* dialog = selectedColor != nullptr ? new QColorDialog(selectedColor, this) : new QColorDialog(defaultColor(), this);
+    dialog->setWindowTitle(tr("Pick color"));
+    dialog->open(this, SLOT(colorSelected(const QColor&)));
```
I think you need to set `Qt::WA_DeleteOnClose` on this one as well. Test by opening a ton of dialogs and closing them, see how the memory use is like.

(FTR some people run Mudlet for months on end without restarting...)

vadi2 commented on src/dlgRoomSymbol.h:
```diff
@@ -0,0 +1,66 @@
+#define MUDLET_DLGROOMSYMBOL_H
+
+/***************************************************************************
+ *   Copyright (C) 2020 by Piotr Wilczynski - delwing@gmail.com            *
```
2021 probably, and in the other file?

SlySven replied:
Given the development period concerned it should probably be `2020-2021`...

vadi2 replied:
Is that really the case, or does only the merge time matter?

vadi2 reviewed (commented):
This is a big piece of work, good job, you can be proud :)

Commit:
remove accidental hints

Commit:
remove not needed variable

Commit:
minor fixes

Commit:
room not exists will not throw error in room char color functions

Commit:
remove properties with auto suggestion

Commit:
remove not neede assignment

Commit:
prevent spawning multiple symbol char dialogs at once

vadi2 reviewed (approved):
This is great :+1:

SlySven commented:
When there is multiple rooms selected with different symbols there is a count missing (it is stuck at `0`) for how many rooms are in the selection:
![image](https://user-images.githubusercontent.com/6163092/103484690-baffa680-4de8-11eb-8a12-57f6dd0612a3.png)

When there are multiple rooms with the SAME symbol the text does not seem to be constructed so that it reports that one is working with more than one room - I don't know if that is a defect from my original coding...!
![image](https://user-images.githubusercontent.com/6163092/103484791-555fea00-4de9-11eb-8353-3891a0734c52.png)

BTW If I set the colour of the rooms for those three death-trap rooms differently then the colour used in the sample is of only one of them:

It isn't obvious on first use that one has to click on the button (?) to change the colour - and the tool tip has a typo (it also needs to be fed the selection size - an `%n` - but I haven't checked the code yet, only the run-time behaviour) - it might be better for it to have some contrasting text on it saying "Change color...":
![image](https://user-images.githubusercontent.com/6163092/103484928-63623a80-4dea-11eb-879b-1bca7489c7b6.png)

SlySven commented:
You might want to consider how you handle the case where only the specified font is to be used and one or more of the symbol glyphs cannot be shown in that font. The two selected before the symbol dialogue was opened are ones that would have the warning triangle in the symbol usage dialogue - but they are not in the map symbol font so if you use it for your sample you won't get anything useful - in the map painting code and the symbol usage code I arranged for that to be handled by inserting the replacement character instead...
![image](https://user-images.githubusercontent.com/6163092/103485182-7bd35480-4dec-11eb-9215-9c70239bc0b5.png)

SlySven commented:
:+1: This *is* an interesting development and it shows potential...

vadi2 commented:
I think that both of us forgot about it shows that this font symbol fallback system is too complicated...

SlySven commented:
It is NOT a fallback system - it is a requirement for a MUD that provides a specific font for its maps - so that the user can stay within that font when they are editing it...

... the fall-back as you describe it is to use all the fonts on the system (which IS the default) but if as a map maker you are using odd glyphs from a particular font (or even a custom one - and I have done stuff in that area myself!) that you are shipping as part of a Mud specific UI you will be trying to see that everyone gets the same map symbols from the same font.

Commit:
fix room symbol selection issues

Delwing commented:
Used parts of logic for creating pixmaps of symbols.
Corrected room counts issues (+ text labels selection)
Fixed default color behavior (after some fixes initial default was always black)

vadi2 commented:
This never made it into the docs 😮
